keps/sig-storage/20190709-csi-snapshot.md

```diff
@@ -0,0 +1,118 @@
+
+### E2E tests
+
+* e2e tests for creating/deleting snapshot.
+* e2e tests for creating volume from snapshot.
+
+## Graduation Criteria
+
+### Alpha->Beta
+
+* Feature complete, including:
+  * Create/delete volume snapshots
+  * Create new volumes from a snapshot
+  * SnapshotContent Deletion/Retain Policy
+  * Snapshot Object in Use Protection
+  * Separate the common controller from the sidecar controller
+  * Add secrets field to list-snapshots RPC in the CSI spec. Add “data-source-secret” in create-volume intended for accessing the data source. Implement them in external-snapshotter and external-provisioner.
+  * Add metrics support
+* Unit and e2e tests implemented
+* Update snapshot CRDs to v1beta1 and enable VolumeSnapshotDataSource feature gate by default.
+
+### Beta->GA
+
+* Snapshot feature is used as a basic building block in other advanced applications. 
+* Feature deployed in production and have gone through at least one K8s upgrade.
+
+## Changes
+
+### Changes Implemented
+
+Here are the changes since the original design proposal:
+
+* Renamed `Ready` to `ReadyToUse` in the `Status` field of `VolumeSnapshot` API object.
+* Changed type of `RestoreSize` in `CSIVolumeSnapshotSource` from `*resource.Quantity`  to `*int64`.
+* Lease based Leader Election support is added.
+* Added `VolumeSnapshotContent` deletion policy which is also specified in `VolumeSnapshotClass`.
+* Added `VolumeSnapshot` and `VolumeSnapshotContent` in Use Protection using Finalizers.
+* Added Finalizer on the snapshot source PVC to prevent it from being deleted when a snapshot is being created from it.
+* Added check to see whether ListSnapshots is supported by the CSI driver. If it is supported, ListSnapshots will be called to find out the status of a snapshot during static binding; otherwise it is assumed the snapshot ID provided by the admin is valid.
+
+### Work in Progress
+
+There are other things we are working on for Beta:
+
+* If snapshot creation times out, VolumeSnapshot status will not be marked as failed so that controller will continue to retry to create until the operation either succeeds or fails. It is up to the user or an upper level application that uses the VolumeSnapshot to determine what to do with the snapshot. This work is on-going.
+* Investigation is on-going to determine whether we should separate common controller logic from other logic that belongs to the sidecar. Can be in the same external-snapshotter repo. The common controller should not be deployed with the driver. It should be deployed by the cluster deployer, or we can provide a way to deploy it as a separate Statefulset, not together with the driver. CRD installation should also be done by the cluster deployer.
+* Add secrets field to list-snapshots RPC in the CSI spec. Add “data-source-secret” in create-volume intended for accessing the data source. Implement them in external-snapshotter and external-provisioner.
+* Add metrics support in the snapshot controller.
```

Comment:
should we make the list of metrics to-be-added more specific? I am thinking about following for BETA:
1. operational end to end latency metrics
    labels: 
       a. operation_name, i.e., creation-snapshot, delete-snapshot
       b. csi-driver-name
2. operation error count
   labels: 
       a. operation_name, i.e., creation-snapshot, delete-snapshot
       b. csi-driver-name

Reply:
Added